resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -338,51 +323,120 @@ private void query(final DnsServerAddressStream nameServerAddrStream,
+                                nameServerAddr, resolvedAddresses, nameServerAddrStream);
+                        query(addressStream, nameServerAddrStreamIndex, question,
+                                queryLifecycleObserver, promise, cause);
                     } else {
-                        // Server did not respond or I/O error occurred; try again.
-                        queryLifecycleObserver.queryFailed(queryCause);
-                        query(nameServerAddrStream, nameServerAddrStreamIndex + 1, question, promise, queryCause);
+                        // Ignore the server and try the next one...
+                        query(nameServerAddrStream, nameServerAddrStreamIndex + 1,
+                                question, queryLifecycleObserver, promise, cause);
                     }
-                } finally {
-                    tryToFinishResolve(nameServerAddrStream, nameServerAddrStreamIndex, question,
-                                       // queryLifecycleObserver has already been terminated at this point so we must
-                                       // not allow it to be terminated again by tryToFinishResolve.
-                                       NoopDnsQueryLifecycleObserver.INSTANCE,
-                                       promise, queryCause);
                 }
+            });
+            if (!DnsNameResolver.doResolveAllCached(nameServerName, additionals, resolverPromise, resolveCache(),
+                    parent.resolvedInternetProtocolFamiliesUnsafe())) {
+                final AuthoritativeDnsServerCache authoritativeDnsServerCache = authoritativeDnsServerCache();
+                new DnsAddressResolveContext(parent, nameServerName, additionals,
+                        parent.newNameServerAddressStream(nameServerName),
+                        resolveCache(), new AuthoritativeDnsServerCache() {
+                    @Override
+                    public List<InetSocketAddress> get(String hostname) {
+                        // To not risk fall into any loop we will not use the cache during follow redirects but only
+                        // on the initial query.
+                        return Collections.emptyList();
+                    }
+
+                    @Override
+                    public void cache(String hostname, InetSocketAddress address, long originalTtl, EventLoop loop) {
+                        authoritativeDnsServerCache.cache(hostname, address, originalTtl, loop);
+                    }
+
+                    @Override
+                    public void clear() {
+                        authoritativeDnsServerCache.clear();
+                    }
+
+                    @Override
+                    public boolean clear(String hostname) {
+                        return authoritativeDnsServerCache.clear(hostname);
+                    }
+                }).resolve(resolverPromise);
             }
-        });
+        } else {
```

Comment:
How about returning at the end of the `if` block above and unwrapping this `else` block?